File: java-compat/Ice/session/SessionFactoryI.java
// **********************************************************************
//
// Copyright (c) 2003-present ZeroC, Inc. All rights reserved.
//
// **********************************************************************

import Demo.*;

class SessionFactoryI extends _SessionFactoryDisp
{
    @Override
    public synchronized SessionPrx create(String name, Ice.Current current)
    {
        SessionI session = new SessionI(name);
        SessionPrx proxy = SessionPrxHelper.uncheckedCast(current.adapter.addWithUUID(session));

        //
        // Remove endpoints to ensure that calls are collocated-only
        // This way, if we invoke on the proxy during shutdown, the invocation fails immediately
        // without attempting to establish any connection
        //
        final SessionPrx collocProxy = SessionPrxHelper.uncheckedCast(proxy.ice_endpoints(new Ice.Endpoint[0]));

        //
        // Never close this connection from the client and turn on heartbeats with a timeout of 30s
        //
        current.con.setACM(Ice.Optional.O(30),
                           Ice.Optional.O(Ice.ACMClose.CloseOff),
                           Ice.Optional.O(Ice.ACMHeartbeat.HeartbeatAlways));
        current.con.setCloseCallback(new Ice.CloseCallback()
                {
                    @Override
                    public void closed(Ice.Connection con)
                    {
                        try
                        {
                            collocProxy.destroy();
                            System.out.println("Cleaned up dead client.");
                        }
                        catch(Ice.LocalException ex)
                        {
                            // The client already destroyed this session, or the server is shutting down
                        }
                    }
                });
        return proxy;
    }

    @Override
    public void shutdown(Ice.Current current)
    {
        System.out.println("Shutting down...");
        current.adapter.getCommunicator().shutdown();
    }
}

File: cpp11/Ice/session/SessionFactoryI.cpp
// **********************************************************************
//
// Copyright (c) 2003-present ZeroC, Inc. All rights reserved.
//
// **********************************************************************

#include <SessionFactoryI.h>
#include <SessionI.h>

using namespace std;
using namespace Demo;

shared_ptr<SessionPrx>
SessionFactoryI::create(string name, const Ice::Current& current)
{
    auto session = make_shared<SessionI>(name);
    auto proxy = Ice::uncheckedCast<SessionPrx>(current.adapter->addWithUUID(session));

    //
    // Remove endpoints to ensure that calls are collocated-only
    // This way, if we invoke on the proxy during shutdown, the invocation fails immediately
    // without attempting to establish any connection
    //
    auto collocProxy = proxy->ice_endpoints(Ice::EndpointSeq());

    //
    // Never close this connection from the client and turn on heartbeats with a timeout of 30s
    //
    current.con->setACM(30, Ice::ACMClose::CloseOff, Ice::ACMHeartbeat::HeartbeatAlways);
    current.con->setCloseCallback([collocProxy](const shared_ptr<Ice::Connection>&)
                                  {
                                      try
                                      {
                                          collocProxy->destroy();
                                          cout << "Cleaned up dead client." << endl;
                                      }
                                      catch(const Ice::LocalException&)
                                      {
                                           // The client already destroyed this session, or the server is shutting down
                                      }
                                  }
        );

    return proxy;
}

void
SessionFactoryI::shutdown(const Ice::Current& current)
{
    cout << "Shutting down..." << endl;
    current.adapter->getCommunicator()->shutdown();
}
